### PR TITLE
rustdoc: general cleanup

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**

### CODE REVIEW

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -218,7 +219,7 @@ struct SharedContext {
 }
 
 impl SharedContext {
-    fn ensure_dir(&self, dst: &Path) -> Result<(), Error> {
+    pub fn ensure_dir(&self, dst: &Path) -> Result<(), Error> {
```
Comment (matklad) [inline src/librustdoc/html/render.rs]: Wouldn't this become an "unreachable pub", once we turn that warning on, due to the fact that `SharedContext` is `crate`?

File: src/librustdoc/html/format.rs [hunk 2]
```diff
@@ -1063,3 +1063,21 @@ impl fmt::Display for DefaultSpace {
         }
     }
 }
+
+pub struct WithFormatter<F>(Cell<Option<F>>);
```
Comment (matklad) [inline src/librustdoc/html/format.rs]: Might go one step further with abstraction:

```
fn display_fn(f: impl FnOnce(&mut fmt::Formatter) -> fmt::Result) -> impl fmt::Display {
    ...
}
```

File: src/librustdoc/html/format.rs [hunk 3]
```diff
@@ -527,21 +521,15 @@ fn tybounds(param_names: &Option<Vec<clean::GenericBound>>) -> impl fmt::Display
...
-    pub fn new(did: DefId, text: &'a str) -> HRef<'a> {
-        HRef { did: did, text: text }
-    }
-}
-
-impl<'a> fmt::Display for HRef<'a> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        if let Some((url, short_ty, fqp)) = href(self.did) {
+pub fn anchor(did: DefId, text: &str) -> impl fmt::Display + '_ {
+    WithFormatter::new(move |f| {
```
Comment (matklad) [inline src/librustdoc/html/format.rs]: I think `WithFormatter` is introduced several commits *ahead*, which is mildly confusing

File: src/librustdoc/html/format.rs (hunk 2 above)
Comment (matklad) [reply src/librustdoc/html/format.rs]: Also, does this need to be `pub` as opposed to just `crate`?

**Comment by GuillaumeGomez** [comment]
I like the direction it's taking: less `Display` is way better indeed.

**Comment by Mark-Simulacrum** [comment]
Okay, I'll fix up @matklad's comments here in the next few days hopefully and probably leave further fmt::Display cleanup to a future PR.

**Comment by Mark-Simulacrum** [comment]
Moved this from description since I both don't want it to get lost but it also isn't noteworthy enough to go into commit history:


I'm still thinking about a good way to deal with the HTML formatting. One possible answer is explored in two of these commits, in particular "Transition CommaSep to function" and "
Replace HRef with anchor function". I'm interested in hearing thoughts on whether their approach looks reasonable, or if we think that there's a better way of moving us off `fmt::Display`. One notable discovery made during this process is that we currently make extensive use of alternate printing to replace `&nbsp;` and such with regular spaces to determine how wide certain things will be when printed; that complicates using normal functions as you still want to thread through a context and then you're not really gaining much over structs.

I can drop any of these commits if we want to discuss them more, but I think most of them here are straight improvements.

**Comment by Mark-Simulacrum** [comment]
Okay, I think I've fixed nits and hopefully CI; this is ready for review.

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors: r+

**Pull request merged**